datameta/api/ui/view.py

```diff
@@ -0,0 +1,194 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sqlalchemy.orm import joinedload, aliased
+from sqlalchemy import func, and_, or_, desc, asc
+
+from pyramid.httpexceptions import HTTPBadRequest, HTTPOk
+from pyramid.request import Request
+from pyramid.view import view_config
+from dataclasses import dataclass
+import webob
+import logging
+import datetime
+
+import pandas as pd
+
+from ... import security, samplesheet, errors
+from ...resource import get_identifier
+from ...models import MetaDatum, MetaDataSet, MetaDatumRecord, User, Group, Submission
+
+from ...api.metadatasets import get_record_from_metadataset, MetaDataSetResponse
+
+from pprint import pprint
+
+log = logging.getLogger(__name__)
+
+@dataclass
+class ViewTableResponse(MetaDataSetResponse):
+    """Data class representing the JSON response returned by POST:/api/ui/view"""
+    submission_label : str = None
+    group_id : dict = None
+
+def metadata_index_to_name(db, idx):
+    name = db.query(MetaDatum.name).order_by(MetaDatum.order).limit(1).offset(idx).scalar()
+    if not name:
+        errors.get_validation_error(["Invalid sort column index"])
+    return name
+
+@view_config(
+    route_name      = "ui_view",
+    renderer        = "json",
+    request_method  = "POST",
+)
+def post(request: Request):
+    """This endpoint implements the datatables API specified at
+    https://datatables.net/manual/server-side
+    to serve the datatables backed table in the "view data" view of the web
+    front end.
+
+    Returns:
+        HTTPFound      200 - if the request could be parsed succesfully
+        HTTPBadRequest 400 - if the request was invalid
+    """
+
+    auth_user   = security.revalidate_user(request)
+    db          = request.dbsession
+
+    # Parse the request as specified by the datatables API.
+    try:
+        # The draw number has to be returned in the response
+        draw     = int(request.POST['draw'])
+        # The number of records to return
+        length   = int(request.POST['length'])
+        # The first record to return, 0 based
+        start    = int(request.POST['start'])
+        # A search pattern the user entered in the search field of the table,
+        # may be empty
+        search   = request.POST['search[value]']
+        # The column index for sorting
+        sort_idx = int(request.POST['order[0][column]'])
+        # The sorting direction
+        sort_asc = request.POST['order[0][dir]'] == 'asc'
```

Comment:
is this caught somewhere that the values can only be `asc` and `desc` or will `anything` also do descending sort?

Reply:
No, that's not caught. Kept it simple here, since we're talking to an API on the other end.